File: src/simulation/agent/behaviors/SwarmBehavior.java
package simulation.agent.behaviors;

import java.util.ArrayList;
import java.util.List;

import simulation.agent.brainPackage.Behavior;
import simulation.common.AVector;
import simulation.common.PolarCoordinate;
import simulation.common.Speed;

public class SwarmBehavior extends Behavior{

boolean limit;	
	
public SwarmBehavior(boolean limit, float l) {
		this.limit = limit;
		SwarmBehavior.l = l;
		alpha = 1 - SwarmBehavior.l;
		setX();
	}

//the larger l, the greater the distance before attraction takes over
private static float l;
private static float alpha;
private float x;

public void setX(){
	x = (float) Math.sqrt(l/alpha);
}

public int generateAngle() {
	int angle = curveCalculation(shiftAngles(0));
	//System.out.println("  ANGLE " + angle);
	//limiting turns between 270 and 90
	if((angle < 270 && angle > 90 && limit)){
		if(angle > 180){
			angle = 270;
		}
		else{
			angle = 90;
		}
	}
	return angle;
}

 
public Speed generateSpeed() {
	return Speed.VERYSLOW;
}

//remove this method
public ArrayList<PolarCoordinate> shiftAngles(int shift){
	ArrayList<PolarCoordinate> p = new ArrayList<PolarCoordinate>();
	
	for(PolarCoordinate deltaPostion : newPostions){
		p.add(deltaPostion);
		int angle = p.get(p.size()-1).getTheta();
		int newAngle = angle - shift; 
		
		if(newAngle < 0){
			angle = 360 - Math.abs(newAngle);
		}
		p.get(p.size()-1).setTheta(angle);
	}
	
	return p;
}

//public for testing only
public int curveCalculation(List<PolarCoordinate> newPostions){
	ArrayList<AVector> vector = new ArrayList<AVector>();
	
	ArrayList<PolarCoordinate> listA = new ArrayList<PolarCoordinate>();
	ArrayList<PolarCoordinate> listR = new ArrayList<PolarCoordinate>();
	
	for(PolarCoordinate p : newPostions){
		if((p.getTheta() > 270 || p.getTheta() < 90 && limit) || !limit){
			if(p.getR() > x && p.getAttraction()){
				listA.add(p);
			}
			else if(p.getAttraction()){
				listR.add(p);
			}
		}
	}
	//System.out.println(listA.size() +  "  " + listR.size() + " total " + (listA.size() + listR.size()));
	
	for(PolarCoordinate p : listA){
		float temp = (float)(p.getR() - Math.sqrt(x));
		float a = (float)((alpha/l)*(Math.pow(temp,2)));
		
		//System.out.println("Curve output " + a);
		
		vector.add(new AVector((float)(a*(Math.sin(Math.toRadians(p.getTheta())))),(float)(a*Math.cos(Math.toRadians(p.getTheta())))));
	}
	
	for(PolarCoordinate p : listR){
		float temp = (float)(-l/(Math.pow(p.getR(), 2)));
		float r = temp + alpha;
		
		vector.add(new AVector((float)(r*(Math.sin(Math.toRadians(p.getTheta())))),(float)(r*Math.cos(Math.toRadians(p.getTheta())))));
	}
	
	return getVectorAngle(vector); 
}

private int getVectorAngle(ArrayList<AVector> vector){
	float iSum = 0;
	float jSum = 0;
	
	for(AVector v : vector){
		iSum = iSum + v.i;
		jSum = jSum + v.j;
	}
	
	//System.out.println(" i " + iSum + "   j " + jSum);
	
	return AVector.getVectorAngle(new AVector(iSum, jSum));
}
}

File: src/simulation/agent/behaviors/SwarmBehavior.ts
import { Behavior } from '../brainPackage/Behavior';
import { PolarCoordinate } from '../../common/PolarCoordinate';
import { AVector } from '../../common/AVector';
import { Speed } from '../../common/Speed';

export class SwarmBehavior extends Behavior {
  private limit: boolean;
  private static l: number = 0;
  private static alpha: number = 0;
  private x: number = 0;

  constructor(limit: boolean, l: number) {
    super();
    this.limit = limit;
    SwarmBehavior.l = l;
    SwarmBehavior.alpha = 1 - SwarmBehavior.l;
    this.setX();
  }

  setX(): void {
    this.x = Math.sqrt(SwarmBehavior.l / SwarmBehavior.alpha);
  }

  generateAngle(): number {
    let angle = this.curveCalculation(this.shiftAngles(0));
    if (angle < 270 && angle > 90 && this.limit) {
      if (angle > 180) {
        angle = 270;
      } else {
        angle = 90;
      }
    }
    return angle;
  }

  generateSpeed(): Speed {
    return Speed.VERYSLOW;
  }

  shiftAngles(shift: number): PolarCoordinate[] {
    const p: PolarCoordinate[] = [];
    for (const deltaPosition of this.newPositions) {
      p.push(deltaPosition);
      let angle = p[p.length - 1].getTheta();
      const newAngle = angle - shift;
      if (newAngle < 0) {
        angle = 360 - Math.abs(newAngle);
      }
      p[p.length - 1].setTheta(angle);
    }
    return p;
  }

  curveCalculation(newPositions: PolarCoordinate[]): number {
    const vector: AVector[] = [];
    const listA: PolarCoordinate[] = [];
    const listR: PolarCoordinate[] = [];
    const l = SwarmBehavior.l;
    const alpha = SwarmBehavior.alpha;
    const x = this.x;

    for (const p of newPositions) {
      // Java: (p.getTheta() > 270 || p.getTheta() < 90 && limit) || !limit
      // && has higher precedence than ||, so: (p.getTheta() > 270 || (p.getTheta() < 90 && limit)) || !limit
      if ((p.getTheta() > 270 || (p.getTheta() < 90 && this.limit)) || !this.limit) {
        if (p.getR() > x && p.getAttraction()) {
          listA.push(p);
        } else if (p.getAttraction()) {
          listR.push(p);
        }
      }
    }

    for (const p of listA) {
      const temp = p.getR() - Math.sqrt(x);
      const a = (alpha / l) * Math.pow(temp, 2);
      vector.push(new AVector(
        a * Math.sin((p.getTheta() * Math.PI) / 180),
        a * Math.cos((p.getTheta() * Math.PI) / 180)
      ));
    }

    for (const p of listR) {
      const temp = -l / Math.pow(p.getR(), 2);
      const r = temp + alpha;
      vector.push(new AVector(
        r * Math.sin((p.getTheta() * Math.PI) / 180),
        r * Math.cos((p.getTheta() * Math.PI) / 180)
      ));
    }

    return this.getVectorAngle(vector);
  }

  private getVectorAngle(vector: AVector[]): number {
    let iSum = 0;
    let jSum = 0;
    for (const v of vector) {
      iSum += v.i;
      jSum += v.j;
    }
    return AVector.getVectorAngle(new AVector(iSum, jSum));
  }
}
